NFC Connection Handover Record and Message library.

KAGA164 opened the pull request:
This PR contains Connection Handover Record an Message library. Additionally it contains NFC initial NFC Pairing sample and encoder and decoder for the Temporary Key value in the OOB LE Record.

To do:
- Support also BLE central role in sample
- TNEP Negotiated Handover
- TNEP Mediated Handover ?
- Update sample na library doc to above

KAGA164 commented:
@grochu @kapi-no Please review

NordicBuilder commented:
**All checks passed.**

## checkpatch (informational only, not a failure)

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:100: WARNING:BLOCK_COMMENT_STYLE: Block comments should align the * on each line
#100: FILE: include/nfc/ndef/ch.h:94:
+	 *    - NFC Forum external type
+	*/

-:223: WARNING:LONG_LINE: line over 80 characters
#223: FILE: include/nfc/ndef/ch.h:217:
+int nfc_ndef_ch_rec_payload_constructor(struct nfc_ndef_ch_rec_payload_desc *ch_rec_payload_desc,

-:368: WARNING:LONG_LINE_COMMENT: line over 80 characters
#368: FILE: include/nfc/ndef/ch.h:362:
+ * NDEF Handover Initiate record descriptor instance.nfc_ndef_ch_hc_rec_payload_desc

-:433: WARNING:LONG_LINE: line over 80 characters
#433: FILE: include/nfc/ndef/ch.h:427:
+int nfc_ndef_ch_hc_rec_payload_constructor(struct nfc_ndef_ch_hc_rec_payload_desc *hc_rec,

-:457: WARNING:LONG_LINE: line over 80 characters
#457: FILE: include/nfc/ndef/ch.h:451:
+#define NFC_NDEF_CH_HC_RECORD_DESC_DEF(_name,				         \

-:458: WARNING:LONG_LINE: line over 80 characters
#458: FILE: include/nfc/ndef/ch.h:452:
+				       _payload_id,			         \

-:459: WARNING:LONG_LINE: line over 80 characters
#459: FILE: include/nfc/ndef/ch.h:453:
+				       _payload_id_length,		         \

-:460: WARNING:LONG_LINE: line over 80 characters
#460: FILE: include/nfc/ndef/ch.h:454:
+				       _payload_desc)			         \

-:461: WARNING:LONG_LINE: line over 80 characters
#461: FILE: include/nfc/ndef/ch.h:455:
+										 \

-:462: WARNING:LONG_LINE: line over 80 characters
#462: FILE: include/nfc/ndef/ch.h:456:
+	NFC_NDEF_GENERIC_RECORD_DESC_DEF(_name,				         \

-:463: WARNING:LONG_LINE: line over 80 characters
#463: FILE: include/nfc/ndef/ch.h:457:
+					 TNF_WELL_KNOWN,                         \

-:464: WARNING:LONG_LINE: line over 80 characters
#464: FILE: include/nfc/ndef/ch.h:458:
+					 _payload_id,                            \

-:465: WARNING:LONG_LINE: line over 80 characters
#465: FILE: include/nfc/ndef/ch.h:459:
+					 _payload_id_length,                     \

-:466: WARNING:LONG_LINE: line over 80 characters
#466: FILE: include/nfc/ndef/ch.h:460:
+					 nfc_ndef_ch_hc_rec_type_field,		 \

-:467: WARNING:LONG_LINE: line over 80 characters
#467: FILE: include/nfc/ndef/ch.h:461:
+					 NFC_NDEF_CH_REC_TYPE_LENGTH,	         \

-:468: WARNING:LONG_LINE: line over 80 characters
#468: FILE: include/nfc/ndef/ch.h:462:
+					 nfc_ndef_ch_hc_rec_payload_constructor, \

-:491: WARNING:LONG_LINE: line over 80 characters
#491: FILE: include/nfc/ndef/ch.h:485:
+int nfc_ndef_ch_ac_rec_payload_constructor(struct nfc_ndef_ch_ac_rec_payload_desc *nfc_rec_ac_payload_desc,

-:513: WARNING:LONG_LINE: line over 80 characters
#513: FILE: include/nfc/ndef/ch.h:507:
+#define NFC_NDEF_CH_AC_RECORD_DESC_DEF(_name,					 \

-:514: WARNING:LONG_LINE: line over 80 characters
#514: FILE: include/nfc/ndef/ch.h:508:
+				       _cps,					 \

-:515: WARNING:LONG_LINE: line over 80 characters
#515: FILE: include/nfc/ndef/ch.h:509:
+				       _carr_data_ref_len,			 \

-:516: WARNING:LONG_LINE: line over 80 characters
#516: FILE: include/nfc/ndef/ch.h:510:
+				       _carr_data_ref,				 \

-:517: WARNING:LONG_LINE: line over 80 characters
#517: FILE: include/nfc/ndef/ch.h:511:
+				       _max_aux_data_ref)			 \

-:518: WARNING:LONG_LINE: line over 80 characters
#518: FILE: include/nfc/ndef/ch.h:512:
+	struct nfc_ndef_ch_ac_rec_data_ref					 \

-:519: WARNING:LONG_LINE: line over 80 characters
#519: FILE: include/nfc/ndef/ch.h:513:
+		_name##_nfc_ac_rec_aux_data_ref_array[_max_aux_data_ref];	 \

-:520: WARNING:LONG_LINE: line over 80 characters
#520: FILE: include/nfc/ndef/ch.h:514:
+	struct nfc_ndef_ch_ac_rec_payload_desc					 \

-:521: WARNING:LONG_LINE: line over 80 characters
#521: FILE: include/nfc/ndef/ch.h:515:
+		_name##_nfc_ac_rec_payload_desc = {				 \

-:522: WARNING:LONG_LINE: line over 80 characters
#522: FILE: include/nfc/ndef/ch.h:516:
+		.cps = _cps,							 \

-:523: WARNING:LONG_LINE: line over 80 characters
#523: FILE: include/nfc/ndef/ch.h:517:
+		.carrier_data_ref = { _carr_data_ref_len, _carr_data_ref },	 \

-:524: WARNING:LONG_LINE: line over 80 characters
#524: FILE: include/nfc/ndef/ch.h:518:
+		.max_aux_data_ref = _max_aux_data_ref,				 \

-:525: WARNING:LONG_LINE: line over 80 characters
#525: FILE: include/nfc/ndef/ch.h:519:
+		.aux_data_ref_count = 0,					 \

-:526: WARNING:LONG_LINE: line over 80 characters
#526: FILE: include/nfc/ndef/ch.h:520:
+		.aux_data_ref = _name##_nfc_ac_rec_aux_data_ref_array		 \

-:527: WARNING:LONG_LINE: line over 80 characters
#527: FILE: include/nfc/ndef/ch.h:521:
+	};									 \

-:528: WARNING:LONG_LINE: line over 80 characters
#528: FILE: include/nfc/ndef/ch.h:522:
+	NFC_NDEF_GENERIC_RECORD_DESC_DEF(_name,					 \

-:529: WARNING:LONG_LINE: line over 80 characters
#529: FILE: include/nfc/ndef/ch.h:523:
+					 TNF_WELL_KNOWN,			 \

-:530: WARNING:LONG_LINE: line over 80 characters
#530: FILE: include/nfc/ndef/ch.h:524:
+					 0,					 \

-:531: WARNING:LONG_LINE: line over 80 characters
#531: FILE: include/nfc/ndef/ch.h:525:
+					 0,					 \

-:532: WARNING:LONG_LINE: line over 80 characters
#532: FILE: include/nfc/ndef/ch.h:526:
+					 nfc_ndef_ch_ac_rec_type_field,		 \

-:533: WARNING:LONG_LINE: line over 80 characters
#533: FILE: include/nfc/ndef/ch.h:527:
+					 NFC_NDEF_CH_REC_TYPE_LENGTH,		 \

-:534: WARNING:LONG_LINE: line over 80 characters
#534: FILE: include/nfc/ndef/ch.h:528:
+					 nfc_ndef_ch_ac_rec_payload_constructor, \

-:552: WARNING:LONG_LINE: line over 80 characters
#552: FILE: include/nfc/ndef/ch.h:546:
+void nfc_ndef_ch_ac_rec_auxiliary_data_ref_clear(struct nfc_ndef_record_desc *ac_rec);

-:565: WARNING:LONG_LINE: line over 80 characters
#565: FILE: include/nfc/ndef/ch.h:559:
+int nfc_ndef_ch_ac_rec_auxiliary_data_ref_add(struct nfc_ndef_record_desc *ac_rec,

-:582: WARNING:LONG_LINE: line over 80 characters
#582: FILE: include/nfc/ndef/ch.h:576:
+int nfc_ndef_ch_cr_rec_payload_constructor(struct nfc_ndef_ch_cr_rec_payload_desc *nfc_rec_cr_payload_desc,

-:601: WARNING:LONG_LINE: line over 80 characters
#601: FILE: include/nfc/ndef/ch.h:595:
+#define NFC_NDEF_CH_CR_RECORD_DESC_DEF(_name,					 \

-:602: WARNING:LONG_LINE: line over 80 characters
#602: FILE: include/nfc/ndef/ch.h:596:
+				       _random_number)				 \

-:603: WARNING:LONG_LINE: line over 80 characters
#603: FILE: include/nfc/ndef/ch.h:597:
+	struct nfc_ndef_ch_cr_rec_payload_desc					 \

-:604: WARNING:LONG_LINE: line over 80 characters
#604: FILE: include/nfc/ndef/ch.h:598:
+		_name##_nfc_cr_rec_payload_desc = {				 \

-:605: WARNING:LONG_LINE: line over 80 characters
#605: FILE: include/nfc/ndef/ch.h:599:
+		.random = _random_number,					 \

-:606: WARNING:LONG_LINE: line over 80 characters
#606: FILE: include/nfc/ndef/ch.h:600:
+	};									 \

-:607: WARNING:LONG_LINE: line over 80 characters
#607: FILE: include/nfc/ndef/ch.h:601:
+	NFC_NDEF_GENERIC_RECORD_DESC_DEF(_name,					 \

-:608: WARNING:LONG_LINE: line over 80 characters
#608: FILE: include/nfc/ndef/ch.h:602:
+					 TNF_WELL_KNOWN,			 \

-:609: WARNING:LONG_LINE: line over 80 characters
#609: FILE: include/nfc/ndef/ch.h:603:
+					 0,					 \

-:610: WARNING:LONG_LINE: line over 80 characters
#610: FILE: include/nfc/ndef/ch.h:604:
+					 0,					 \

-:611: WARNING:LONG_LINE: line over 80 characters
#611: FILE: include/nfc/ndef/ch.h:605:
+					 nfc_ndef_ch_cr_rec_type_field,		 \

-:612: WARNING:LONG_LINE: line over 80 characters
#612: FILE: include/nfc/ndef/ch.h:606:
+					 NFC_NDEF_CH_REC_TYPE_LENGTH,		 \

-:613: WARNING:LONG_LINE: line over 80 characters
#613: FILE: include/nfc/ndef/ch.h:607:
+					 nfc_ndef_ch_cr_rec_payload_constructor, \

-:616: WARNING:LONG_LINE_COMMENT: line over 80 characters
#616: FILE: include/nfc/ndef/ch.h:610:
+/** @brief Macro for accessing the NFC NDEF Alternative Carrier record descriptor

-:750: WARNING:LONG_LINE: line over 80 characters
#750: FILE: include/nfc/ndef/pair_msg.h:42:
+int nfc_ndef_pair_msg_le_oob_encode(struct nfc_ndef_le_oob_rec_payload_desc *oob,

-:1176: WARNING:LONG_LINE: line over 80 characters
#1176: FILE: samples/bluetooth/nfc_pairing/src/main.c:101:
+					   &NFC_NDEF_LE_OOB_RECORD_DESC(oob_rec),

-:1585: WARNING:LONG_LINE: line over 80 characters
#1585: FILE: subsys/nfc/ndef/ch.c:49:
+static u32_t nfc_ac_rec_payload_size_get(const struct nfc_ndef_ch_ac_rec_payload_desc *ac_rec_payload_desc)

-:1606: WARNING:LONG_LINE: line over 80 characters
#1606: FILE: subsys/nfc/ndef/ch.c:70:
+static u32_t nfc_hc_payload_size_get(const struct nfc_ndef_ch_hc_rec_payload_desc *hc_rec)

-:1618: WARNING:LONG_LINE: line over 80 characters
#1618: FILE: subsys/nfc/ndef/ch.c:82:
+int nfc_ndef_ch_rec_payload_constructor(struct nfc_ndef_ch_rec_payload_desc *ch_rec_payload_desc,

-:1624: WARNING:LONG_LINE_COMMENT: line over 80 characters
#1624: FILE: subsys/nfc/ndef/ch.c:88:
+		/* There must be at least 1 free byte in buffer for version byte. */

-:1655: WARNING:LONG_LINE: line over 80 characters
#1655: FILE: subsys/nfc/ndef/ch.c:119:
+		(struct nfc_ndef_ch_rec_payload_desc *)ch_rec->payload_descriptor;

-:1664: WARNING:LONG_LINE: line over 80 characters
#1664: FILE: subsys/nfc/ndef/ch.c:128:
+		(struct nfc_ndef_ch_rec_payload_desc *)ch_rec->payload_descriptor;

-:1669: WARNING:LONG_LINE: line over 80 characters
#1669: FILE: subsys/nfc/ndef/ch.c:133:
+int nfc_ndef_ch_hc_rec_payload_constructor(struct nfc_ndef_ch_hc_rec_payload_desc *hc_rec,

-:1685: WARNING:SUSPECT_CODE_INDENT: suspect code indent for conditional statements (8, 20)
#1685: FILE: subsys/nfc/ndef/ch.c:149:
+	if ((hc_rec->ctf < TNF_WELL_KNOWN) ||
[...]
+		    return -EFAULT;

-:1687: WARNING:TABSTOP: Statements should start on a tabstop
#1687: FILE: subsys/nfc/ndef/ch.c:151:
+		    return -EFAULT;

-:1709: WARNING:LONG_LINE: line over 80 characters
#1709: FILE: subsys/nfc/ndef/ch.c:173:
+int nfc_ndef_ch_ac_rec_payload_constructor(struct nfc_ndef_ch_ac_rec_payload_desc *nfc_rec_ac_payload_desc,

-:1712: WARNING:LONG_LINE: line over 80 characters
#1712: FILE: subsys/nfc/ndef/ch.c:176:
+	u32_t payload_size = nfc_ac_rec_payload_size_get(nfc_rec_ac_payload_desc);

-:1746: WARNING:LONG_LINE: line over 80 characters
#1746: FILE: subsys/nfc/ndef/ch.c:210:
+	for (size_t i = 0; i < nfc_rec_ac_payload_desc->aux_data_ref_count; i++) {

-:1762: WARNING:LONG_LINE: line over 80 characters
#1762: FILE: subsys/nfc/ndef/ch.c:226:
+void nfc_ndef_ch_ac_rec_auxiliary_data_ref_clear(struct nfc_ndef_record_desc *ac_rec)

-:1765: WARNING:LONG_LINE: line over 80 characters
#1765: FILE: subsys/nfc/ndef/ch.c:229:
+		(struct nfc_ndef_ch_ac_rec_payload_desc *)ac_rec->payload_descriptor;

-:1770: WARNING:LONG_LINE: line over 80 characters
#1770: FILE: subsys/nfc/ndef/ch.c:234:
+int nfc_ndef_ch_ac_rec_auxiliary_data_ref_add(struct nfc_ndef_record_desc *ac_rec,

-:1774: WARNING:LONG_LINE: line over 80 characters
#1774: FILE: subsys/nfc/ndef/ch.c:238:
+		(struct nfc_ndef_ch_ac_rec_payload_desc *)ac_rec->payload_descriptor;

-:1776: WARNING:LONG_LINE: line over 80 characters
#1776: FILE: subsys/nfc/ndef/ch.c:240:
+	if (ac_rec_payload->aux_data_ref_count >= ac_rec_payload->max_aux_data_ref) {

-:1780: WARNING:LONG_LINE: line over 80 characters
#1780: FILE: subsys/nfc/ndef/ch.c:244:
+	ac_rec_payload->aux_data_ref[ac_rec_payload->aux_data_ref_count].data = aux_data;

-:1781: WARNING:LONG_LINE: line over 80 characters
#1781: FILE: subsys/nfc/ndef/ch.c:245:
+	ac_rec_payload->aux_data_ref[ac_rec_payload->aux_data_ref_count].length = aux_length;

-:1787: WARNING:LONG_LINE: line over 80 characters
#1787: FILE: subsys/nfc/ndef/ch.c:251:
+int nfc_ndef_ch_cr_rec_payload_constructor(struct nfc_ndef_ch_cr_rec_payload_desc *nfc_rec_cr_payload_desc,

-:1893: WARNING:LONG_LINE: line over 80 characters
#1893: FILE: subsys/nfc/ndef/le_oob_rec_parser.c:220:
+								    NFC_NDEF_LE_OOB_REC_TK_LEN);

-:1974: WARNING:LONG_LINE: line over 80 characters
#1974: FILE: subsys/nfc/ndef/pair_msg.c:55:
+int nfc_ndef_pair_msg_le_oob_encode(struct nfc_ndef_le_oob_rec_payload_desc *oob,

-:2042: WARNING:LONG_LINE: line over 80 characters
#2042: FILE: subsys/nfc/ndef/pair_msg.c:123:
+					       &NFC_NDEF_CH_AC_RECORD_DESC(cr_rec));

- total: 0 errors, 81 warnings, 2011 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.

```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

joerchan commented:
@KAGA164 Could you please fix the checkpatch warnings?

ru-fu commented on include/nfc/ndef/ch.rst:
```diff
@@ -0,0 +1,30 @@
+The Connection Handover defines the NDEF records and the corresponding message structure that supports negotiation and activation of an alternative communication carrier.
+The negotiated communication carrier can then be used to perform certain activities between the two devices, such as Bluetooth pairing.
+The Connection Hanover module provides functions for creating and encoding Handover Connection records.
+
```
I agree - sample code is really helpful.

ru-fu commented on include/nfc/ndef/pair_msg.rst:
```diff
@@ -0,0 +1,25 @@
+.. _nfc_pair:
+
+NFC Pair messages
```
Is "NFC Pair message" a term? I couldn't find it anywhere.
If not, maybe we should combine this RST file with the ch.rst file into "Connection handover messages and records", as we do for URI and generic messages/records. That would require the files to be named more similarly though so it is clear that those two libraries are meant to be used together.

KAGA164 replied:
I think we can merge documentation into one file and provide a code sample there. Yes this libraries in most cases will be used together

KAGA164 replied:
What about changing name `pair_msg` -> `ch_msg`

KAGA164 commented on subsys/nfc/ndef/le_oob_rec.c:
```diff
@@ -84,6 +85,22 @@ static int le_role_encode(enum nfc_ndef_le_oob_rec_le_role le_role, u8_t **buff,
 	return 0;
 }
 
+int nfc_ndef_le_oob_rec_tk_generate(u8_t *buff)
```
I removed all random number generation form Connection Handover a pair message library. Now this should be done on the application level

KAGA164 commented on subsys/nfc/ndef/le_oob_rec_parser.c:
```diff
@@ -361,6 +384,12 @@ void nfc_ndef_le_oob_rec_printout(
+	if (le_oob_rec_desc->tk_value) {
+		LOG_HEXDUMP_INF(le_oob_rec_desc->tk_value,
+				NFC_NDEF_LE_OOB_REC_TK_LEN,
+				"\tTk Value:");
```
Fixed

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,623 @@
+ * record, defined in the file @c ch_rec.c. It is used in the
+ * @ref NFC_NDEF_HS_RECORD_DESC_DEF macro.
+ */
+extern const u8_t nfc_ndef_ch_hs_rec_type_field[];
```
All type field definition moved to `payload_type_common`

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,623 @@
+ * record, defined in the file @c ch_rec.c. It is used in the
+ * @ref NFC_NDEF_HS_RECORD_DESC_DEF macro.
+ */
+extern const u8_t nfc_ndef_ch_hs_rec_type_field[];
```
Ok

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,623 @@
+		/** Optionall data length. */
+		u8_t data_length;
+
+		/** Optionall information about alternative carrier.
```
Fixed

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,623 @@
+		/** Optionall data length. */
+		u8_t data_length;
+
+		/** Optionall information about alternative carrier.
```
Fixed

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,623 @@
+		const u8_t *type;
+
+		/** Optionall data length. */
+		u8_t data_length;
```
Ok shortened

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,623 @@
+	/** Number of Auxiliary Data Reference fields. */
+	u8_t aux_data_ref_count;
+
+	/** Number of Auxiliary Data Reference fields. */
```
Fixed

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,623 @@
+	enum nfc_ndef_ch_ac_rec_cps cps;
+
+	/** Carrier Data Reference. */
+	struct nfc_ndef_ch_ac_rec_data_ref carrier_data_ref;
```
Removed

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,623 @@
+	struct nfc_ndef_ch_ac_rec_data_ref carrier_data_ref;
+
+	/** Maximum number of Auxiliary Data Reference fields. */
+	const u8_t max_aux_data_ref;
```
Changed

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,623 @@
+/**
+ * @brief Alternative Carrier record payload descriptor.
+ */
+struct nfc_ndef_ch_ac_rec_payload_desc {
```
Removed

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,623 @@
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_rec_payload_constructor(struct nfc_ndef_ch_rec_payload_desc *ch_rec_payload_desc,
```
Changed

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,623 @@
+/**
+ * @brief Handover Select record payload descriptor.
+ */
+struct nfc_ndef_ch_rec_payload_desc {
```
Ok :)

KAGA164 commented on subsys/nfc/ndef/ch.c:
```diff
@@ -0,0 +1,268 @@
+}
+
+int nfc_ndef_ch_ac_rec_payload_constructor(struct nfc_ndef_ch_ac_rec_payload_desc *nfc_rec_ac_payload_desc,
+					   u8_t *buff, u32_t *len)
```
Changed

KAGA164 commented on include/nfc/ndef/pair_msg.h:
```diff
@@ -0,0 +1,153 @@
+ *
+ * This function encodes an NFC NDEF message into a buffer.
+ *
+ * @param[in] obb Pointer to the LE OOB Record payload descriptor.
```
Fixed

KAGA164 commented on include/nfc/ndef/pair_msg.h:
```diff
@@ -0,0 +1,153 @@
+ * Connection Hanover Select message. Number of the Alternative
+ * Carrier Records must be equal to the number of the Carrier Records.
+ *
+ * @param[in] obb Pointer to the Alternative Carrier Records.
```
Fixed

KAGA164 commented:
@ru-fu I squashed your commit but also I renamed `pair_msg` to `ch_msg` and merge their doc  into one file. Please check it again.
@joerchan I get-ride with the most of checkpatch warnings, I left only few with to long line when changing may reduce readability

grochu commented on subsys/nfc/ndef/ch_msg.c:
```diff
@@ -0,0 +1,177 @@
+	int err;
+
+	NFC_NDEF_MSG_DEF(hm_msg, cnt + 1);
+	NFC_NDEF_CH_HS_RECORD_DESC_DEF(hm_rec, NFC_CH_MAJOR_VER,
```
```suggestion
	NFC_NDEF_CH_HM_RECORD_DESC_DEF(hm_rec, NFC_CH_MAJOR_VER,
```

grochu replied:
or unify these functions, and use a common macro if unified according to a previous comment

KAGA164 replied:
Fixed

grochu commented on subsys/nfc/ndef/ch_msg.c:
```diff
@@ -0,0 +1,177 @@
+	int err;
+
+	NFC_NDEF_MSG_DEF(hi_msg, cnt + 1);
+	NFC_NDEF_CH_HS_RECORD_DESC_DEF(hi_rec, NFC_CH_MAJOR_VER,
```
```suggestion
	NFC_NDEF_CH_HI_RECORD_DESC_DEF(hi_rec, NFC_CH_MAJOR_VER,
```

grochu replied:
or unify these functions, and use a common macro if unified according to a previous comment

KAGA164 replied:
Fiexed

grochu commented on samples/bluetooth/nfc_pairing/src/main.c:
```diff
@@ -0,0 +1,477 @@
+
+	bt_addr_le_t addr;
+
+	if (!k_msgq_get(&bonds_queue, &addr, K_NO_WAIT)) {
```
If `CONFIG_BT_MAX_PAIRED > 1` we will start directed advertising to the first peer we are bonded with?

KAGA164 replied:
Yes we started advertising to first bonded peer which is not connected. When dir adv failed or there is empty connection slot, advertising will be started to next bonded peer.

grochu commented on samples/bluetooth/nfc_pairing/src/main.c:
```diff
@@ -0,0 +1,477 @@
+	err = bt_le_oob_set_sc_data(conn, &oob_local.le_sc_data, NULL);
+	if (err) {
+		printk("Error while setting OOB data: %d\n", err);
+	} else {
```
I think this is not needed, user will know if the pairing is successful. If setting OOB returns an error, user will also be notified by the line above.

KAGA164 replied:
Yes right. Removed

grochu commented on samples/bluetooth/nfc_pairing/src/main.c:
```diff
@@ -0,0 +1,477 @@
+		return;
+	}
+
+	conn_cnt++;
```
shouldn't we call `bt_conn_ref()` here?

KAGA164 replied:
No, we want to count connection not refering existing one

grochu replied:
OK, makes sense

grochu commented on samples/bluetooth/nfc_pairing/src/main.c:
```diff
@@ -0,0 +1,477 @@
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	conn_cnt--;
```
`bt_conn_unref()` ?

KAGA164 replied:
No beacause we do not create reference to given connection we only count them

grochu commented on include/nfc/ndef/ch_msg.h:
```diff
@@ -0,0 +1,155 @@
+ * Carrier Records must be equal to the number of the Carrier Records.
+ *
+ * @param[in] ac Pointer to the Alternative Carrier Records.
+ * @param[in] carrier Pointer to the Carrier Records which identifies
```
Can this be clarified? According to the code, this argument can be a pointer to element 0 of an array of Carrier Record descriptors. This applies to the series of functions _Hs, Hm, Hr, Hi_

KAGA164 replied:
I added check for NULL pointer and count according to CH specification records without Alternative Carrier and Carrier record does not make a lot of sense.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/nrfconnect/sdk-nrf?pullRequest=2205) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/nrfconnect/sdk-nrf?pullRequest=2205) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: KAGA164<br/>:x: ru-fu<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/nrfconnect/sdk-nrf?pullRequest=2205) it.</sub>

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+		u8_t data_len;
+
+		/** Optional information about alternative carrier.
+		 *  It length can be calculated using following formula:
```
Done

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+ *
+ * @param[in] ch_rec Pointer to the Connection Handover record
+ *                   description.
+ * @param[in] buf Pointer to the payload destination. If NULL, function will
```
Done

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_rec_payload_encode(struct nfc_ndef_ch_rec *ch_rec,
```
Done

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+ *
+ * @param[in] _name Name of the created record descriptor instance.
+ * @param[in] _rec_type_field The NFC Forum Well Known Type
+ *                            Handover Connection record type.
```
Done

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+ * @brief Function for clearing local records in the NFC NDEF Connection
+ * Handover records.
+ *
+ * This function clears local records from the Connection Handover records.
```
Done

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+ *                           Connection Handover specification.
+ * @param[in] _max_record Maximum number of local records.
+ */
+#define NFC_NDEF_CH_GENERIC_RECORD_DESC_DEF(_name,			 \
```
Yes good idea. Generic term will not be used anymore

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+ *         Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_rec_local_record_add(struct nfc_ndef_record_desc *ch_rec,
+				     struct nfc_ndef_record_desc *local_rec);
```
Done

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+ *
+ * @param[in] hc_rec Pointer to the Connection Handover Carrier record
+ *                   description.
+ * @param[in] buf Pointer to the payload destination. If NULL, function will
```
Done

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_hc_rec_payload_encode(struct nfc_ndef_ch_hc_rec *hc_rec,
```
Done

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+ *                           Connection Handover specification.
+ * @param[in] _max_record Maximum number of local records.
+ */
+#define NFC_NDEF_CH_HC_RECORD_DESC_DEF(_name,				    \
```
This is `Handover Carrier record`

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+ *         Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_ac_rec_auxiliary_data_ref_add(struct nfc_ndef_record_desc *ac_rec,
+					      u8_t *aux_data, u8_t aux_length);
```
Done

KAGA164 commented on include/nfc/ndef/ch.h:
```diff
@@ -0,0 +1,572 @@
+ *
+ * @param[in] nfc_rec_cr Pointer to the Collision Resolution
+ *                       record description.
+ * @param[in] buf Pointer to the payload destination. If NULL, function will
```
Done

KAGA164 commented on include/nfc/ndef/ch_msg.h:
```diff
@@ -0,0 +1,155 @@
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_msg_le_oob_encode(struct nfc_ndef_le_oob_rec_payload_desc *oob,
```
Done

KAGA164 commented on include/nfc/ndef/ch_msg.h:
```diff
@@ -0,0 +1,155 @@
+
+/**@file
+ *
+ * @defgroup nfc_ndef_ch_msg NFC Connection Hanover pairing messages
```
Ok removed

KAGA164 commented on include/nfc/ndef/ch_msg.h:
```diff
@@ -0,0 +1,155 @@
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_msg_hs_encode(struct nfc_ndef_record_desc *ac,
```
Done

KAGA164 commented on include/nfc/ndef/ch_msg.h:
```diff
@@ -0,0 +1,155 @@
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_msg_hr_encode(struct nfc_ndef_record_desc *ac,
```
Done

KAGA164 commented on include/nfc/ndef/ch_msg.h:
```diff
@@ -0,0 +1,155 @@
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_msg_hs_encode(struct nfc_ndef_record_desc *ac,
+			      struct nfc_ndef_record_desc *carrier,
```
Done

KAGA164 commented on include/nfc/ndef/ch_msg.h:
```diff
@@ -0,0 +1,155 @@
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_msg_hr_encode(struct nfc_ndef_record_desc *ac,
+			      struct nfc_ndef_record_desc *carrier,
```
Done

KAGA164 commented on include/nfc/ndef/ch_msg.h:
```diff
@@ -0,0 +1,155 @@
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_msg_hm_encode(struct nfc_ndef_record_desc *ac,
```
Done

KAGA164 commented on include/nfc/ndef/ch_msg.h:
```diff
@@ -0,0 +1,155 @@
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_msg_hm_encode(struct nfc_ndef_record_desc *ac,
+			      struct nfc_ndef_record_desc *carrier,
```
Done

KAGA164 commented on include/nfc/ndef/ch_msg.h:
```diff
@@ -0,0 +1,155 @@
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_msg_hi_encode(struct nfc_ndef_record_desc *ac,
```
Done

KAGA164 commented on include/nfc/ndef/ch_msg.h:
```diff
@@ -0,0 +1,155 @@
+ *           Otherwise, a (negative) error code is returned.
+ */
+int nfc_ndef_ch_msg_hi_encode(struct nfc_ndef_record_desc *ac,
+			      struct nfc_ndef_record_desc *carrier,
```
Done

KAGA164 commented on samples/bluetooth/nfc_pairing/README.rst:
```diff
@@ -0,0 +1,85 @@
+Bluetooth: NFC pairing
+######################
+
+The NFC pairing sample demonstrates Bluetooth LE Connection Handover pairing using an NFC tag.
```
Done

KAGA164 commented on samples/bluetooth/nfc_pairing/README.rst:
```diff
@@ -0,0 +1,85 @@
+The message that the tag sends to the NFC device contains data required to initiate pairing.
+To start the NFC data transfer, the NFC device must touch the NFC antenna that is connected to the development kit.
+
+After reading the tag, the device can pair with the device that is advertising.
```
Done

KAGA164 commented on samples/bluetooth/nfc_pairing/README.rst:
```diff
@@ -0,0 +1,85 @@
+######################
+
+The NFC pairing sample demonstrates Bluetooth LE Connection Handover pairing using an NFC tag.
+It can be used to test the touch-to-pair feature between Nordic Semiconductor's devices and an NFC reader or another NFC polling device, for example, a mobile phone.
```
Done

KAGA164 commented on samples/bluetooth/nfc_pairing/README.rst:
```diff
@@ -0,0 +1,85 @@
+It can be used to test the touch-to-pair feature between Nordic Semiconductor's devices and an NFC reader or another NFC polling device, for example, a mobile phone.
+
+The sample shows the usage of NFC NDEF :ref:`nfc_ch`.
+It uses minimal Bluetooth functionality and implements only the Device Information Service.
```
Done

KAGA164 commented on samples/bluetooth/nfc_pairing/README.rst:
```diff
@@ -0,0 +1,85 @@
+After programming the sample to your board, test it by performing the following steps:
+
+1. Touch the NFC antenna with the smartphone or tablet and observe that LED 2 is lit.
+#. Confirm pairing with ``Nordic_NFC_pairing`` in a pop-up windows on the smartphone or tablet and observe that LED 1 lights up.
```
Done

KAGA164 commented on include/nfc/ndef/ch_msg.rst:
```diff
@@ -0,0 +1,74 @@
+****************************
+
+Connection Handover messages contain at least one Connection Handover record.
+They are used to establish an NFC connection and pair with other Bluetooth devices.
```
Done

KAGA164 commented on include/nfc/ndef/ch_msg.rst:
```diff
@@ -0,0 +1,74 @@
+
+The Connection Handover message library provides functions for encoding the following messages:
+
+* Generic OOB message
```
We can 👍  Changed

KAGA164 commented on include/nfc/ndef/ch_msg.rst:
```diff
@@ -0,0 +1,74 @@
+Connection Handover messages
+****************************
+
+Connection Handover messages contain at least one Connection Handover record.
```
Yes, removed

KAGA164 commented on include/nfc/ndef/ch_msg.rst:
```diff
@@ -0,0 +1,74 @@
+
+The Connection Handover message library provides functions for encoding the following messages:
+
+* Generic OOB message
```
Changed

KAGA164 commented on include/nfc/ndef/le_oob_rec.h:
```diff
@@ -26,6 +26,9 @@
 #endif
 
+/** Temporary Key length. */
+#define NFC_NDEF_LE_OOB_REC_TK_LEN 16
```
It is not defined anywhere. There is only mention how long this key should be.

KAGA164 commented:
@kapi-no @grochu Please check again. I think that unification of CH message can be done during work on the OOB TNEP
